### DISCUSSION TITLE
Broken pipe or closed connection

### DISCUSSION THREAD

**Discussion by vol4ikman** [php-amqplib/php-amqplib#1073] [Help wanted]
After I am trying to receive the message from the server I got this error. I receive the response after 4 mins.....

The error after send/receive the message:

`<b>Fatal error</b>:  Uncaught PhpAmqpLib\Exception\AMQPConnectionClosedException: Broken pipe or closed connection in /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/StreamIO.php:171
Stack trace:
#0 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/AMQPIOReader.php(62): PhpAmqpLib\Wire\IO\StreamIO-&gt;read()
#1 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/AMQPReader.php(35): PhpAmqpLib\Wire\AMQPIOReader-&gt;rawread()
#2 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(601): PhpAmqpLib\Wire\AMQPReader-&gt;read()
#3 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(658): PhpAmqpLib\Connection\AbstractConnection-&gt;wait_frame()
#4 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php(235): PhpAmqpLib\Connection\AbstractConnection-&gt;wait_channel()
#5 /home/so in <b>/home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/StreamIO.php</b> on line <b>171</b><br />`

**Comment by vol4ikman** [comment]
`require_once __DIR__ . '/lib/vendor/autoload.php';
	require 'lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AMQPStreamConnection.php';
	require 'lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AMQPSSLConnection.php';
	require 'lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Message/AMQPMessage.php';

	define( 'CERTS_PATH', realpath( __DIR__ . '/lib/' ) );

	$response = array(
		'message_sent'     => false,
		'message'          => '',
		'received-message' => '',
	);

	$ssl_opts = array(
		'cafile'           => CERTS_PATH . '/ca-cert.pem',
		'local_cert'       => CERTS_PATH . '/client-cert.pem',
		'local_pk'         => CERTS_PATH . '/client-key.pem',
		'verify_peer'      => false,
		'verify_peer_name' => false,
	);

	$host     = 'XXXX';
	$user     = 'XXX';
	$password = 'XXXX';
	$port     = 12021; // secured.

	// Send message.
	$connection = new PhpAmqpLib\Connection\AMQPSSLConnection( $host, $port, $user, $password, $vhost = '/', $ssl_opts );
	// $connection   = new PhpAmqpLib\Connection\AMQPStreamConnection( $host, $port, $user, $password );

	$channel = $connection->channel();

	$channel->queue_declare( 'request-queue', false, true, false, false );

	$message_content = '{"type": "user_information","meter_id": "73505064136"}';

	$msg = new PhpAmqpLib\Message\AMQPMessage( $message_content );

	$channel->basic_publish( $msg, '', 'request-queue' );

	$response['message']      = "[x] Sent 'user_information json' to request-queue.";
	$response['message_sent'] = true;

	$channel->close();
	$connection->close();

	// Receive the message.
	$connection = new PhpAmqpLib\Connection\AMQPSSLConnection( $host, $port, $user, $password, $vhost = '/', $ssl_opts );
	$channel    = $connection->channel();
	$channel->queue_declare( 'reply-queue', false, true, false, false );

	$received_message = '';

	$callback = function ( $reply_msg ) {
		// print_r( $reply_msg->body );.
		$received_message = $reply_msg->body;
	};
	$channel->basic_consume( 'reply-queue', '', false, true, false, false, $callback );

	while ( $channel->is_open() ) {
		$channel->wait();
	}

	$response['received-message'] = $received_message;
	$channel->close();
	$connection->close();

	print_r( $response );
	die();`

**Comment by ramunasd** [comment]
Any logs from broker side? I guess connection was closed due to inactivity. What is default connection timeout?
You don't use heartbeats nor keepalive. So it's a matter of time when connection will be closed if subscribed queue is not full.

Can You try with heartbeats enabled, higher timeouts and `$channel->consume()` instead of endless loop?

**Comment by ramunasd** [comment]
BTW, You publish to one queue but consume from another. How it supposed to work? Looks like You need to start from learning how RabbitMQ works in general.
https://www.rabbitmq.com/tutorials/tutorial-three-php.html

**Comment by vol4ikman** [comment]
When I am trying to create the connection according to example you sent 

"$connection = new PhpAmqpLib\Connection\AMQPStreamConnection( $host, $port, $user, $password );"

I got this error:

`<b>Fatal error</b>:  Uncaught PhpAmqpLib\Exception\AMQPInvalidFrameException: Invalid frame type 21 in /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php:606
Stack trace:
#0 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(658): PhpAmqpLib\Connection\AbstractConnection-&gt;wait_frame()
#1 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php(235): PhpAmqpLib\Connection\AbstractConnection-&gt;wait_channel()
#2 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php(352): PhpAmqpLib\Channel\AbstractChannel-&gt;next_frame()
#3 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(266): PhpAmqpLib\Channel\AbstractChannel-&gt;wait()
#4 /home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(237): PhpAmqpLib in <b>/home/solegreen/public_html/alex/lib/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php</b> on line <b>606</b><br />`
